File: examples/gimbal_set_mode.cpp
#include "stdio.h"
#include <pthread.h>
#include <cstdlib>
#include <string>
using namespace std;

#include"payloadSdkInterface.h"

#if (CONTROL_METHOD == CONTROL_UART)
T_ConnInfo s_conn = {
    CONTROL_UART,
    payload_uart_port,
    payload_uart_baud
};
#else
T_ConnInfo s_conn = {
    CONTROL_UDP,
    udp_ip_target,
    udp_port_target
};
#endif

PayloadSdkInterface* my_payload = nullptr;
bool is_exit = false;

void quit_handler(int sig);
void onPayloadParamChanged(int event, char* param_char, double* param_double);

int main(int argc, char *argv[]){
	printf("Starting Set gimbal mode example...\n");
	signal(SIGINT,quit_handler);

	// create payloadsdk object
	my_payload = new PayloadSdkInterface(s_conn);

	// register callback function
	my_payload->regPayloadParamChanged(onPayloadParamChanged);

	// init payload
	my_payload->sdkInitConnection();
	printf("Waiting for payload signal! \n");

	my_payload->checkPayloadConnection();
	
#if 0
	/** From the software v307.54
		- We will no longer user the command_long for setting the gimbal's mode
		- We will use the flag in the message MAVLINK_MSG_ID_GIMBAL_DEVICE_SET_ATTITUDE (284) for the gimbal mode directly
		- The camera definition file will be changed to remove all option for gimbal's mode
	**/

	printf("Gimbal set mode LOCK, delay in 5 secs \n");
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_LOCK, PARAM_TYPE_UINT32);
	usleep(5000000);

	printf("Gimbal set mode FOLLOW, delay in 5 secs \n");
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_FOLLOW, PARAM_TYPE_UINT32);
	usleep(5000000);

	printf("Gimbal set mode MAPPING, delay in 5 secs \n");
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_MAPPING, PARAM_TYPE_UINT32);
	usleep(5000000);

	printf("Gimbal set mode OFF, delay in 5 secs \n");
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_OFF, PARAM_TYPE_UINT32);
	usleep(5000000);

	printf("Gimbal set mode RESET, delay in 5 secs \n");
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, PAYLOAD_CAMERA_GIMBAL_MODE_RESET, PARAM_TYPE_UINT32);
	usleep(5000000);
#else
	uint16_t _flags = 0;
	
	// Need to get the gimbal status flags
	printf("Check the current gimbal status flags \n");
	_flags = my_payload->getGimbalDeviceStatusFlags();

	while(!is_exit){
		printf("Change gimbal mode to LOCK \n");
		_flags |= GIMBAL_DEVICE_FLAGS_YAW_LOCK;
		_flags &= ~(1 << 14);					// xoa co
		_flags &= ~GIMBAL_DEVICE_FLAGS_NEUTRAL;	// xoa co
		_flags &= ~GIMBAL_DEVICE_FLAGS_RETRACT; // xoa co
		my_payload->setGimbalMode(_flags);
		usleep(3000000);

		printf("Change gimbal mode to FOLLOW \n");
		_flags &= ~GIMBAL_DEVICE_FLAGS_YAW_LOCK;
		_flags &= ~(1 << 14);					// xoa co
		_flags &= ~GIMBAL_DEVICE_FLAGS_NEUTRAL;	// xoa co
		_flags &= ~GIMBAL_DEVICE_FLAGS_RETRACT; // xoa co
		my_payload->setGimbalMode(_flags);
		usleep(3000000);

		printf("Change gimbal power to OFF \n");
		_flags |= GIMBAL_DEVICE_FLAGS_RETRACT;
		_flags &= ~(1 << 14);					// xoa co
		_flags &= ~GIMBAL_DEVICE_FLAGS_NEUTRAL;	// xoa co
		my_payload->setGimbalMode(_flags);
		usleep(3000000);

		printf("Change gimbal power to ON \n");
		_flags &= ~GIMBAL_DEVICE_FLAGS_RETRACT;
		_flags &= ~(1 << 14);					// xoa co
		_flags &= ~GIMBAL_DEVICE_FLAGS_NEUTRAL;	// xoa co
		my_payload->setGimbalMode(_flags);
		usleep(8000000);

		printf("Move gimbal to 20 degrees yaw \n");
		my_payload->setGimbalSpeed(0, 0 , 20, INPUT_ANGLE);
		usleep(3000000);

		printf("Set gimbal return HOME \n");
		_flags |= GIMBAL_DEVICE_FLAGS_NEUTRAL;
		_flags &= ~GIMBAL_DEVICE_FLAGS_RETRACT; // xoa co
		_flags &= ~(1 << 14);					// xoa co mapping
		my_payload->setGimbalMode(_flags);
		usleep(100000);
		// reset the control bit
		_flags &= ~GIMBAL_DEVICE_FLAGS_NEUTRAL;	// xoa co
		my_payload->setGimbalMode(_flags);
		usleep(3000000);

		printf("Set gimbal to Mapping mode \n");
		_flags |= (1 << 14);
		_flags &= ~GIMBAL_DEVICE_FLAGS_NEUTRAL;	// xoa co
		_flags &= ~GIMBAL_DEVICE_FLAGS_RETRACT; // xoa co
		my_payload->setGimbalMode(_flags);
		usleep(3000000);

		printf("Set gimbal return HOME \n");
		_flags |= GIMBAL_DEVICE_FLAGS_NEUTRAL;
		_flags &= ~GIMBAL_DEVICE_FLAGS_RETRACT; // xoa co
		_flags &= ~(1 << 14);					// xoa co mapping
		my_payload->setGimbalMode(_flags);
		usleep(100000);
		// reset the HOME bit
		_flags &= ~GIMBAL_DEVICE_FLAGS_NEUTRAL;	// xoa co
		my_payload->setGimbalMode(_flags);
		usleep(3000000);
	}

#endif

	while(true){
		// main loop
		usleep(10000000);
	}

	// close payload interface
	try {
		my_payload->sdkQuit();

	}
	catch (int error){}

	return 0;
}

void quit_handler( int sig ){
    printf("\n");
    printf("TERMINATING AT USER REQUEST \n");
    printf("\n");

    is_exit = true;
    // close payload interface
    try {
        my_payload->sdkQuit();
    }
    catch (int error){}

    // end program here
    exit(0);
}

void onPayloadParamChanged(int event, char* param_char, double* param){
	switch(event){
	case PAYLOAD_GB_ATTITUDE:{
		// param[0]: param_index
		// param[1]: value
		printf(" --> Gimbal Mode: %s\n", param_char);
		break;
	}
	default: break;
	}
}

File: gimbal_set_mode.py
import time
import signal
import sys
from pymavlink import mavutil
from libs.payload_sdk import PayloadSdkInterface
from libs.payload_define import *

my_payload = None

# Signal handler for quitting
def quit_handler(sig, frame):
    global my_payload
    print("\nTERMINATING AT USER REQUEST")

    # Close payload interface
    if my_payload:
        try:
            my_payload.sdkQuit()
        except Exception as e:
            print(f"Error while quitting payload: {e}")
    # End program    
    sys.exit(0)

def main():
    global my_payload

    print("Starting Set gimbal mode example...\n")
    signal.signal(signal.SIGINT, quit_handler)

    # Create payloadsdk object
    my_payload = PayloadSdkInterface()

    # Init payload
    my_payload.sdkInitConnection()
    print("Waiting for payload signal!\n")

    # Check connection
    my_payload.checkPayloadConnection()

    # Set gimbal to LOCK mode
    print("Gimbal set LOCK mode, delay in 5 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_LOCK, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(5) 

    # Set gimbal to FOLLOW mode
    print("Gimbal set FOLLOW mode, delay in 5 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_FOLLOW, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(5)

    # Set gimbal to MAPPING mode
    print("Gimbal set MAPPING mode, delay in 5 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_MAPPING, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(5)

    # Set gimbal to OFF mode
    print("Gimbal set OFF mode, delay in 5 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_OFF, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(5)

    # Set gimbal to RESET mode
    print("Gimbal set RESET mode, delay in 5 secs")
    my_payload.setPayloadCameraParam(PAYLOAD_CAMERA_GIMBAL_MODE, payload_camera_gimbal_mode.PAYLOAD_CAMERA_GIMBAL_MODE_RESET, mavutil.mavlink.MAV_PARAM_TYPE_UINT32)
    time.sleep(5)

    # Close payload interface
    try:
        my_payload.sdkQuit()
    except Exception as e:
        print(f"Error while quitting payload: {e}")

if __name__ == "__main__":
    main()